### ISSUE TITLE
Dump all raw measurements

### ISSUE THREAD

**Issue by vincenthz** [vincenthz/hs-gauge#4]
Each rounds of measurement in a benchmarks could optionally be dumped in CSV format

```csv
name-bench,clock-time-diff,cpu-time-diff,cycles-diff,gc-diff...
...
```

It should make measurement/analysis debugging easier, plus it could be loaded or transformed without re-running the measurement.

**Comment by harendra-kumar** [comment]
We can even use `show` and `read` for serialization the way I used it for the `--measure-only` switch. That switch in fact is a raw-dump to file but it dumps only one benchmark. That way we won't need code to manage yet another serialization to csv.

**Comment by jberryman** [comment]
I use my own tool to compare and analyze criterion runs. If you could clone the CSV output functionality of criterion, I would happily start using `hs-gauge`

**Comment by harendra-kumar** [comment]
@jberryman sure, it should be easy. BTW, is the tool you are using reusable for wider audience?

**Comment by jberryman** [comment]
It's a really messy tweaked version of a tool kindly shared by bgamari here https://github.com/jberryman/criterion-compare . There are probably some projects that are better and do more, but I haven't looked into it.

But it also occurs to me: I think criterion should move the web frontend report generation to a standalone tool that I only need to install and update occasionally (assuming that's a big source of the dependencies, which I'm not sure is the case)

**Comment by vincenthz** [comment]
@jberryman your plan sounds very similar to what I have in mind; Along with what is available right now in gauge (displaying in terminal), I'm planning to record all the measurement and move all the analysis+display to a (probably) purescript based program.

**Comment by jberryman** [comment]
Yeah in fiddling with that tool I posted I did realize that doing the statistics in JS if possible is really the way to go, since it lets you be really flexible in your UI.

This is OT but is this a fork of criterion? You describe it as a "clone" so I'm assuming a re-implementation. Do you track the criterion bug reports (presumably some of which might have ended up in here, depending on how you went about starting this project)? Also I noticed what looked like a bunch of work on the html reporting stuff in criterion; it might be worth looking at that.

For me my priorities in terms of visualization/reports are:
- being able to compare across runs in a rigorous and intuitive way (see at a glance if a modification is a net improvement), and track regressions
- being able to dynamically compare benchmarks in a single run: filter viewing of the results with a search string (since graphs become useless unless all benchmarks are approx same scale), "invert" groupings so that we can toggle between e.g. comparing reads and writes within libraries X and Y respectively, and comparing libraries `Y` and `Z` with each other (i.e. comparing performance of both `read` implementations).

**Comment by vincenthz** [comment]
yes it's a fork of criterion, not a clone. I don't think I've described it as a clone anywhere, but I could be wrong.

I don't follow criterion bug reports personally, but the codebase is starting to diverge significantly specially after the rounds of @harendra-kumar 's refactor, cleanup and improvements. So if there's anything that get fixed there, it won't be fixed here.

As to the HTML display, it was removed completely from gauge early, and it's very unlikely that it will be put back in most form. I think this separate program for display that is more dynamic is the right way to go... apart from the fact that someone need to write it :).

I think there's a bunch of "hidden factors" in criterion that could be dynamically adjusted in a dynamic UI to refine what the user is actually looking at.
